Webumenia 1053 make items sortable within topic

igor-kamil opened the pull request:
add functionality to reorder and remove items within topic in laravel backpack
- added custom field `items_sortable`
- reused already existing `jquery-ui sortable`
- possibility to drag-and-drop items or set order manually

eronisko commented on resources/views/vendor/backpack/crud/fields/items_sortable.blade.php:
```diff
@@ -0,0 +1,99 @@
+    <label>{!! $field['label'] !!}</label>
+
+    <table class="table table-sm table-striped" id="table">
+        {{-- <p>Connected items:</p> --}}
```
```suggestion
```

eronisko commented on resources/views/vendor/backpack/crud/fields/items_sortable.blade.php:
```diff
@@ -0,0 +1,99 @@
+<!-- field_type_name -->
+<div @include('crud::inc.field_wrapper_attributes') >
+    <label>{!! $field['label'] !!}</label>
```
Just curious, do we need `{!!` here?

igor-kamil replied:
Probably. They use it in their own components: `/vendor/backpack/crud/src/resources/views/fields/*` and is also in the template in documentation https://backpackforlaravel.com/docs/3.6/crud-fields#creating-a-custom-field-type

But to be honest - I don't know. Personally I wouldn't use it.

eronisko commented on resources/views/vendor/backpack/crud/fields/items_sortable.blade.php:
```diff
@@ -0,0 +1,99 @@
+        </tbody>
+    </table>
+
+    {{-- HINT --}}
```
IMHO these comments don't add much value

igor-kamil replied:
I am just overriding the template ¯\_(ツ)_/¯ (btw - they kept it also in their templates :D )

eronisko commented on resources/views/vendor/backpack/crud/fields/items_sortable.blade.php:
```diff
@@ -0,0 +1,99 @@
+             }).disableSelection();
+
+             $('input.sorter').on('keypress', function (e) {
+                 if(e.which === 13){
```
This might be a good place for that named constant we talked about :) (I had to google for that key code)
Alternatively, I'd leave it out altogether and save the Enter key for its default purpose -- submitting the whole form.

igor-kamil replied:
you are right. fixed

eronisko commented on resources/views/vendor/backpack/crud/fields/items_sortable.blade.php:
```diff
@@ -0,0 +1,99 @@
+                    {{ $item->type  }}
+                </td>
+                <td>
+                    <input class="sorter form-control" size="2" data-id="{{ $item->id }}" value="{{ $item->pivot->order }}" name="items_sort[{{ $item->id }}]">
```
I'd probably drop the input field and just leave re-ordering by dragging.

igor-kamil replied:
@Endut requested it. he will use it.

eronisko commented:
@igor-kamil this is not terribly important but I've noticed a bug. When I try to replace first or last item, it doesn't always work. To reproduce, make 3 items, and try replacing item 1 with item 2 several times.

eronisko commented:
@igor-kamil I've also noticed another thing. When you change the order, the browser makes 54 requests in the background. They're all pretty fast (they all seem to hit cache) but it's strange nevertheless 🤷‍♂ I've tested on Chrome.

igor-kamil commented:
@eronisko with the bug - you are right. but it appears only after repeating the reordering multiple times only for first and last item. 
I removed the `containment` option and added `axis: "y"` to limit movement. and now it works properly even when reordering multiple times ;) 

regarding the 54 requests... that's crazy... but nothing what I can handle 🤷‍♂. good to know.
